engine/kv_engine.hpp

```diff
@@ -205,14 +205,134 @@ class KVEngine : public Engine {
     *collection_ptr = (CollectionType*)hash_entry.GetIndex().ptr;
 
     // check collection is expired.
-    if (TimeUtils::CheckIsExpired((*collection_ptr)->GetExpiredTime())) {
+    if (TimeUtils::CheckIsExpired((*collection_ptr)->ExpireTime())) {
       hash_table_->Erase(entry_ptr);
       // TODO(Zhichen): add background cleaner.
       return Status::NotFound;
     }
     return s;
   }
 
+  struct LookupResult {
+    Status s{Status::NotSupported};
+    HashEntry entry{};
+    HashEntry* entry_ptr{nullptr};
+  };
+
+  // lookupKeyImpl does not check for deleted or expired key.
+  // It returns whatever it finds as is.
+  // It's up to caller to handle the deleted or expired key.
+  template <bool readonly>
+  LookupResult lookupKeyImpl(StringView key) {
+    LookupResult result;
+    HashTable::KeyHashHint hint = hash_table_->GetHint(key);
+    if (readonly) {
+      result.s =
+          hash_table_->SearchForRead(hint, key, PrimaryRecordType,
+                                     &result.entry_ptr, &result.entry, nullptr);
+    } else {
+      result.s = hash_table_->SearchForWrite(hint, key, PrimaryRecordType,
+                                             &result.entry_ptr, &result.entry,
+                                             nullptr);
+    }
+    return result;
+  }
+
+  // Look up the key, checks for expiration or deletion.
+  // If key is deleted or has expired, handle it to ExpireCleaner
+  // Also checks whether the type matches expected type.
+  template <bool owns_lock>
+  LookupResult lookupKey(StringView key, RecordType expected_type) {
+    LookupResult result = lookupKeyImpl<true>(key);
+    if (result.s != Status::Ok) {
+      kvdk_assert(result.s == Status::NotFound, "");
+      return result;
+    }
+    if (result.entry.GetRecordType() == RecordType::StringDeleteRecord) {
```

Comment:
You may get SortedDeleteRecord here too, use mask GetRecordType() & DeleteRecordType